### PR TITLE
gh-106004: Add PyDict_GetItemRef() function

### PR DESCRIPTION
**Pull request by vstinner**
* Add PyDict_GetItemRef() and PyDict_GetItemStringRef() functions.
* Add unit tests on the PyDict C API in test_capi.

* Issue: gh-106004

----
:books: Documentation preview :books:: https://cpython-previews--106005.org.readthedocs.build/

### CODE REVIEW

File: Doc/c-api/dict.rst
Comment (erlend-aasland) [inline Doc/c-api/dict.rst]:
```suggestion
   * If the key is present, set *\*value* to a new :term:`strong reference`
     to the value and return ``0``.
   * If the key is missing, set *\*value* to ``NULL`` and return ``0``.
```

Comment (erlend-aasland) [reply Doc/c-api/dict.rst]: Oops, never mind.

File: Doc/c-api/dict.rst
Comment (erlend-aasland) [inline Doc/c-api/dict.rst]:
```suggestion
.. c:function:: int PyDict_GetItemRef(PyObject *p, PyObject *key, PyObject **pvalue)
```

File: Objects/dictobject.c
Comment (erlend-aasland) [inline Objects/dictobject.c]: Why not strongly typed, since it is a `PyDict_` API?

```suggestion
PyDict_GetItemRef(PyDictObject *op, PyObject *key, PyObject **pvalue)
```

**Comment by markshannon** [comment]
What's the rationale for not distinguishing between found and not found in the return value?
See https://github.com/python/devguide/issues/1121

**Comment by markshannon** [comment]
Can we come up with a better name than `PyDict_GetItemRef`?
I see why you are adding `Ref` to the end, but *all* API functions should return new references, so it is a bit like calling the function `PyDict_GetItemNotWrong`.

Obviously, the ideal name is already taken.
Anyone have any suggestions for a better name?

**Comment by vstinner** [comment]
> What's the rationale for not distinguishing between found and not found in the return value?

Sure, if you think that it's useful, I can return 1 if the key is present.

It doesn't go against the syntax ``if (PyDict_GetItemRef(...) < 0) ...`` to check for error. I dislike having to use 2 variables to check for a function variable, so I prefer to use the implicit test in an ``if (...)`` statement. Pseudo-code with a variable to store the function result and dispatch the action depending on it:

```
PyObject *value;
int res = PyDict_GetItemRef(dict, key, &value);
if (res < 0) ... error ...
else if (res == 0) ... missing key ...
else ... present key
```

Well, I'm fine with people having different taste (if it doesn't go against my taste ;-)).

> Can we come up with a better name than PyDict_GetItemRef?

Nope, that's my local optimum name :-)

> Anyone have any suggestions for a better name?

Maybe PyObject_GetItem()? :-D Just kidding, PyObject_GetItem() already exists and works for any type (doesn't crash if it's not a dict) :-)

Well, that would just be 2 more functions on the top of the existing 9 functions to "get an item in dict":

* PyDict_GetItem()
* PyDict_GetItemString()
* PyDict_GetItemWithError()
* _PyDict_GetItemIdWithError()
* _PyDict_GetItemStringWithError()
* _PyDict_GetItemWithError()
* _PyDict_GetItem_KnownHash()

Ok, let me propose a name: PyDict_GetItemOkThisTimeItShouldBeCorrect() :-)

I already added the following functions with the "Ref" suffix:

* PyModule_AddObjectRef()
* PyImport_AddModuleRef()
* PyWeakref_GetRef()

The "Ref" is here to remain that this one it returns a strong **ref**erence.

And also: Py_NewRef() and Py_XNewRef() (not really a "strong reference" variant of an existing function).

**Comment by markshannon** [comment]
Testing the result is always at least as efficient as doing a test on `*pvalue`.
You need to handle all three cases, so there needs to be two tests regardless. 

We made https://github.com/python/devguide/issues/1121 to avoid having to repeat this discussion for every new API function.
If you think returning 0 for both found and not-found is better, can you explain why on that issue. Thanks.

**Comment by markshannon** [comment]
> I already added the following functions with the "Ref" suffix:
> 
> * PyModule_AddObjectRef()
> * PyImport_AddModuleRef()
> * PyWeakref_GetRef()

Can we please stop adding new functions to the C-API until we have established conventions.
Otherwise, we will need to add yet another variant that conforms to the naming convention.

**Comment by vstinner** [comment]
I modified the API to return 1 when the key is present (and 0 when the key is missing).

**Comment by vstinner** [comment]
> Can we please stop adding new functions to the C-API until we have established conventions. Otherwise, we will need to add yet another variant that conforms to the naming convention.

I created https://github.com/capi-workgroup/problems/issues/52 to discuss C API naming convention (when adding new functions).

File: Objects/dictobject.c
Comment (vstinner) [reply Objects/dictobject.c]: I would like to move the whole C API towards the only usage of ``PyObject*``. Currently, some functions expect a specific type like PyDictObject or PyTupleObject, whereas some others expect PyObject.

Moreover, existing functions use PyObject.

Comment (erlend-aasland) [reply Objects/dictobject.c]: I created python/devguide#1127 to try to find a consensus regarding guidelines for weakly/stronly typed APIs.

**Comment by vstinner** [comment]
@erlend-aasland @markshannon: Are you ok with the proposed API? @erlend-aasland added some [guidelines related to proposed API in the Devguide](https://devguide.python.org/developer-workflow/c-api/index.html#guidelines-for-expanding-changing-the-public-api).

@erlend-aasland @markshannon: About the function name, I created https://github.com/capi-workgroup/problems/issues/52 but it's unclear to me what's the outcome of this naming convention discussion.

What should I do to make progress on this PR?

File: Objects/dictobject.c
Comment (erlend-aasland) [reply Objects/dictobject.c]: AFAICS, this is the only unresolved issue; there are pros and cons for each variant (`PyObject *` vs. `PyDictObject *`). I'm not sure we can find consensus in the near future :(

**Comment by markshannon** [comment]
No, I'm not OK with the proposed API. I've already said that. 

Using `PyObject *` is needlessly throwing away type information, and we haven't established naming conventions yet.
Let's agree on the rules first, before adding more functions that will need to be changed to the API.

What's the rush?

**Comment by markshannon** [comment]
TBH, I don't think you should have opened this PR until there was some feedback on the issue.

File: Objects/dictobject.c
Comment (vstinner) [reply Objects/dictobject.c]: If the argument type is changed to ``PyDictObject*``, it means that all existing calls to PyDict_GetItem() and PyDict_GetItemWithError() will also have to cast their argument to ``PyDictObject*`` when the code is updated to use ``PyDict_GetItemRef()``.

Just one example: ``PyDict_New()`` returns a ``PyObject*``. It sounds wrong to me that the caller has more knowledge about the expected type than the official ``PyDict_New()`` API.

```c
PyObject *obj = PyDict_New():
PyDictObject *dict_for_getitemref = (PyDictObject*)obj;  // why??
(...)
PyObject *value;
if (PyDict_GetItemStringRef(dict_for_getitemref, "key", &value) < 0) ... errror ...
```

The whole C API of PyDict requires ``PyObject*`` type: PyDict_Size(), PyDict_Update(), PyDict_Values(), PyDict_SetDefault(), ...

**If** the C API would be rewritten from scratch, maybe it would make sense to use different types. But this PR is an **incremental fix** of the Python C API. Changing the parameter type would better fit into the idea of a brand new C API: https://github.com/capi-workgroup/problems/issues/55

**Comment by vstinner** [comment]
By the way, draft [PEP 703 – Making the Global Interpreter Lock Optional in CPython](https://peps.python.org/pep-0703/) proposes ``PyDict_FetchItem()`` which returns a strong reference, to replace PyDict_GetItem() (which returns a borrowed reference). The PEP has a section explaining why PyDict_GetItem() is **not** deprecated: [Why Not Deprecate PyDict_GetItem in Favor of PyDict_FetchItem?](https://peps.python.org/pep-0703/#why-not-deprecate-pydict-getitem-in-favor-of-pydict-fetchitem). I understand that the proposed API is: ``PyObject* PyDict_FetchItem(PyObject *dict, PyObject *key)``.

**Comment by vstinner** [comment]
> Using PyObject * is needlessly throwing away type information,

I replied there: https://github.com/python/cpython/pull/106005#discussion_r1243884378

> we haven't established naming conventions yet.

I was confused by @gpshead's [comment](https://github.com/capi-workgroup/problems/issues/52#issuecomment-1608393695): "I don't think it matters which naming scheme we pick or even if it winds up wholly consistent."

> What's the rush?

There is no rush. I just wanted to understand the status of this PR.

**Review by erlend-aasland — APPROVED** [review]
I'm approving this. A new naming scheme makes sense for a new API; I'm not sure it makes sense to try and enforce a new scheme in the current API. For now, there is already precedence of the `Ref` suffix in the current API; I'm ok with that. Also, the current API uses `PyObject *` all over the place. If we are to change this, we practically will end up with a completely new API; AFAICS, there is no problem with sticking to the current practice.

**Review by gpshead — APPROVED** [review]

**Comment by vstinner** [comment]
I rebased my PR on the main branch to fix a merge conflict.

@markshannon: Are you against this API? My PR got two approvals. I would like to know your opinion about it :-)

**Comment by vstinner** [comment]
@serhiy-storchaka (on the issue):
> The only problem is that functions with so similar names have completely different interface. It is pretty confusing. Would not be better to name it PyDict_LookupItem or like? It may be worth to add also PyMapping_LookupItem for convenience. BTW, I think it is time to make _PyObject_LookupAttr public (if we agree about the name).

As I explained earlier, IMO it's better to have a similar name to ease the migration and to ease the discovery of the newer "correct" API. @erlend-aasland and @gpshead are fine with it.

For me, "Get" is the obvious name. For example, the Python API uses ``__getitem__()`` method name, the Python dict type has ``get()`` and ``setdefault()`` methods.

**Comment by vstinner** [comment]
@serhiy-storchaka (on the issue):
> The only problem is that functions with so similar names have completely different interface. It is pretty confusing. Would not be better to name it PyDict_LookupItem or like? It may be worth to add also PyMapping_LookupItem for convenience. BTW, I think it is time to make _PyObject_LookupAttr public (if we agree about the name).

As I explained earlier, IMO it's better to have a similar name to ease the migration and to ease the discovery of the newer "correct" API. @erlend-aasland and @gpshead are fine with it.

For me, "Get" is the obvious name. For example, the Python API uses ``__getitem__()`` method name, the Python dict type has ``get()`` and ``setdefault()`` methods.

**Comment by serhiy-storchaka** [comment]
The "Ref" suffix is not informative. In different functions it have different meaning (`Py_IncRef()`, `PyWeakref_CheckRef()` -- nothing in common). But comparing with `PyModule_AddObjectRef()` and `PyImport_AddModuleRef()` I would expect that `PyDict_GetItemRef()` has the same signature as `PyDict_GetItem()` -- takes two `PyObject*` arguments, returns `PyObject*`, the only difference is in refcounts here or there. It is not, and you should completely rewrite the code which uses it.

BTW, `PyDict_GetItem()` is not directly related to `__getitem__`. It does not call `__getitem__`, and `__getitem__` does not call it.

**Comment by vstinner** [comment]
@serhiy-storchaka: Did you see https://github.com/capi-workgroup/problems/issues/52 discussion?

**Comment by serhiy-storchaka** [comment]
I seen it, but have not found a link to it at a time. Thank you for the link. The discussions are scattered over so many places.

**Comment by vstinner** [comment]
@gpshead created https://github.com/python/steering-council/issues/201 to ask the Steering Council opinion about PyDict_GetItemRef(), and adding similar APIs (to avoid borrowed references) in general.

**Comment by vstinner** [comment]
So this issue is now waiting for the Steering Council? Apparently, @markshannon is blocking this change, whereas @erlend-aasland and @gpshead approved it.

Well, i don't know if it's taken in account, but i'm in favor of my change 😅

**Comment by vstinner** [comment]
@markshannon:
> What's the rationale for not distinguishing between found and not found in the return value?
See https://github.com/python/devguide/issues/1121

PyDict_GetItemRef() now returns 1 if the item is found, and 0 if the item is not found.

By the way, I recently modified PyWeakref_GetRef() to return 1 if the reference is alive, and return 0 if it's dead.

**Comment by vstinner** [comment]
@colesbury @ambv: PEP 703 proposes PyDict_FetchItem(). What do you think of this proposed API?

**Comment by vstinner** [comment]
@markshannon raised different questions.

Change the first parameter type to ``PyDictObject*`` and use assertions rather than runtime checks on the object type. I propose to keep ``PyObject*`` to make the new API closer to existing APIs, knowing that most PyDict API use ``PyObject*``, especially PyDict_New(). -- I'm open to reconsider this. See https://github.com/capi-workgroup/problems/issues/31 and https://github.com/python/devguide/issues/1127 onging discussions.

Concern about the lack of naming convention when adding new functions to the C API: I created https://github.com/capi-workgroup/problems/issues/52 for that.

@markshannon: would you approve this change if I change the parameter type to ``PyDictObject*`` znd convert runtime type check to an assertion?

File: Doc/c-api/dict.rst
Comment (serhiy-storchaka) [inline Doc/c-api/dict.rst]: In other functions I used `obj` and `result` instead of `p` and `pvalue`.

File: Objects/dictobject.c
Comment (serhiy-storchaka) [inline Objects/dictobject.c]: It can be further simplified:
```c
    result = PyDict_GetItemRef((PyObject *)dv->dv_dict, key, &found);
    if (result > 0) {
        result = PyObject_RichCompareBool(found, value, Py_EQ);
        Py_DECREF(found);
    }
    return result;
```

**Comment by markshannon** [comment]
> Apparently, @markshannon is blocking this change, whereas @erlend-aasland and @gpshead approved it.

I'm opposed because making ad-hoc changes like this is going to make the C-API worse, not better.
Why not engage in the discussion at https://github.com/capi-workgroup first, then we won't need this debate for *every* PR that changes the C-API?

@erlend-aasland @gpshead, I'm curious why you prefer this sort of ad-hoc change to a more principled approach?

**Review by serhiy-storchaka — APPROVED** [review]
It would prefer if the name be similar to the name of abstract API function with which it shares signature and behavior, `PyMapping_GetOptionalItem()`, but existing `PyDict_GetItem()` and `PyDict_GetItemWithError()` functions already return an optional item, the difference with them is that the new function will return a strong reference. So I have no ideas about the name and remove myself from this discussion.

In general, I support adding this function. The benefits:
* Returns a strong reference. It will save from some errors and may be better for PyPy.
* Save CPU time for calling PyErr Occurred().

**Comment by vstinner** [comment]
> Why not engage in the discussion at https://github.com/capi-workgroup first, then we won't need this debate for every PR that changes the C-API?

From my point of view, the Problems project is sometimes too theorical, and working on a way more concrete API helps me to see how discussions apply to "real code".

This PR is connected to multiple Problems issues, references in comments above.

File: Doc/c-api/dict.rst
Comment (vstinner) [reply Doc/c-api/dict.rst]: It seems like the whole PyDict documentation uses *p* for the first parameter. If you want to change it, please open a separated PR to change all of them.

For pvalue, hum, PyDict_Next() uses *pkey* and *pvalue* parameter names... But Python is not always consistent :-( The header files uses *key* and *value* names. The C implementation again uses *pkey* and *pvalue* names. I checked other functions which have a ``PyObject**``: only a small minority uses a ``p`` prefix to show that it's a pointer to an object. So I renamed the parameter to **result**: thanks, good idea.

File: Objects/dictobject.c
Comment (vstinner) [reply Objects/dictobject.c]: Oh nice, my first implementation didn't return 1 when the item was found.

I took your code with a minor nit: ``if (result == 1)`` instead of ``if (result > 0)``.

**Comment by vstinner** [comment]
I rebased my PR to fix a merge conflict, I renamed "pvalue" parameter to "result", and I simplified dictitems_contains() implementation.

**Comment by colesbury** [comment]
In the context of PEP 703, I think it would be better to have variations that *only* change one axis of the semantics (e.g., new vs. borrowed, error vs. no error) and have the naming reflect that. For example, PEP 703 proposes:

`PyDict_FetchItem` for `PyDict_GetItem` and
`PyDict_FetchItemWIthError` for `PyDict_GetItemWithError`.

I don't care too much about the specific names - what I'm getting at is that I think it's important to have variants that return "new references" for all the existing functions that return "borrowed references" and the naming convention should allow that.

**Comment by vstinner** [comment]
> I think it would be better to have variations that only change one axis of the semantics (e.g., new vs. borrowed, error vs. no error) 

We are discussing C API guidelines to prevent repeating old design mistakes when adding new functions. Returning -1 on error rather than NULL is now required by these new guidelines: https://devguide.python.org/developer-workflow/c-api/index.html#guidelines-for-expanding-changing-the-public-api

Yes, migrating from PyDict_GetItemWithError() to PyDict_GetItemRef() is non-trivial, but there are multiple benefits of using the newer API:

* Return a strong reference, instead of a borrowed reference
* Avoid having to check "expensive" PyErr_Occurred() to check if an error was raising: the function result directly provide the information
* If you want, you can use the result to check if the key is present or not (alternative to checking if *value* is ``NULL``).

**Comment by vstinner** [comment]
> @markshannon: would you approve this change if I change the parameter type to PyDictObject* and convert runtime type check to an assertion?

I implemented this change at the 3rd commit:

* Change first argument type to ``PyDictObject*``
* Replace runtime type check with an assertion: ``assert(PyDict_Check(mp));``
* Remove C API tests on passing the wrong type as the first PyDict_GetItemRef() argument since this case does now crash, instead of raising SystemError and returning -1.
* Implementation detail: I had to declare the **PyTypeObject** type name in the limited C API for this name.

File: Objects/dictobject.c
Comment (serhiy-storchaka) [inline Objects/dictobject.c]: `pvalue` or `result`?

File: Objects/dictobject.c
Comment (serhiy-storchaka) [inline Objects/dictobject.c:5199]: Or maybe even `value != NULL`?

**Review by serhiy-storchaka — COMMENTED** [review]
I am dislike using concrete struct types instead of `PyObject*` in API, especially in public API. Isn't there a rule forbidding this?

**Comment by vstinner** [comment]
> I am dislike using concrete struct types instead of PyObject* in API, especially in public API. Isn't there a rule forbidding this?

There is a discussion to change this: https://github.com/capi-workgroup/problems/issues/31 and https://github.com/python/devguide/issues/1127. It sounds like a good idea to avoid a runtime type check by expecting a specific type. I wrote it in the What's New in Python 3.13 entry of this PR:

> The first argument type is not checked at runtime for best performance, instead the argument type is ``PyDictObject*``.

File: Objects/dictobject.c
Comment (vstinner) [reply Objects/dictobject.c]: Oh, I missed that comment. I updated the comment and moved the comment to the header file to make it more visible (more people read header files, than people reading C files).

File: Objects/dictobject.c
Comment (vstinner) [reply Objects/dictobject.c:5199]: Well, I adapted your suggested ``(result > 0)`` code.

Testing ``result == 1`` and ``value != NULL`` is exactly the same: if it's not the case, it's a big bug in the function implementation. Here since we store the function result in a variable, I like to reuse it.

Micro-optimization note. Also, maybe comparing a "small" int is faster than comparing a "large" pointer :-) Especially because function result is usually stored in a common "fast" CPU register :-)

**Comment by serhiy-storchaka** [comment]
It is not about runtime checking. It is about requiring to cast the argument to `PyDictObject*` every time you use the function: `PyDict_GetItemRef((PyDictObject*)foo, bar, &baz)`. It is tiresome, and it is unsafe, because the compiler will not reject the code if `foo` is `int` or `const char*`.

**Comment by vstinner** [comment]
> It is not about runtime checking. It is about requiring to cast the argument to PyDictObject* every time you use the function: PyDict_GetItemRef((PyDictObject*)foo, bar, &baz). It is tiresome, and it is unsafe, because the compiler will not reject the code if foo is int or const char*.

Please join the discussion at https://github.com/capi-workgroup/problems/issues/31 where these points were already raised and discussed.

**Comment by gpshead** [comment]
Our C API *only* accepts plain `PyObject *` as input to all our public APIs. Otherwise user code will be littered with typecasts all over the place.

**Review by gpshead — CHANGES_REQUESTED** [review]
commenting here to tell github to remove my approval as this PR seems to be going in exploratory directions and is awaiting various feedback and decisions.

**Comment by AlexWaygood** [comment]
> commenting here to tell github to remove my approval as this PR seems to be going in exploratory directions and is awaiting various feedback and decisions.

FYI for the future -- you can "dismiss" a prior review to remove your approval without submitting a new review:

image

**Comment by vstinner** [comment]
> commenting here to tell github to remove my approval as this PR seems to be going in exploratory directions and is awaiting various feedback and decisions.

This PR is ready for review, so I remove the draft status again.

**Comment by vstinner** [comment]
> @serhiy-storchaka @gpshead @erlend-aasland: Are you ok with the API using PyDictObject* type? Or should I revert this 3rd commit?

Oh. @serhiy-storchaka and @gpshead seem to be strongly against it, so I reverted the change to go back to the initial plan: ``PyObject*`` type.

I tried to use a revert, but sadly the Git history became too complicated, so instead of squashed commits and I rebased my PR. Sorry about that :-(

---

Gregory:
> commenting here to tell github to remove my approval as this PR seems to be going in exploratory directions and is awaiting various feedback and decisions.

Well, I was requested to set *multiple* guidelines for different aspects of the API on this single function addition.

* Return 1 if the key is present?
  
  * My first version returned 0 if the key is present and if the key is missing, both considered as "success".
  * A guideline was added to the devguide: https://github.com/python/devguide/issues/1121
  * I modified my PR to return 1 if the key is present (and return 0 if the key is missing).

* Function name

  * My PR adds **Ref** suffix to the existing function
  * @markshannon dislikes the name and asked to define a naming convention for new functions.
  * I created: "Naming convention" https://github.com/capi-workgroup/problems/issues/52
  * @serhiy-storchaka proposed PyDict_GetOptionalItem() but is ok with PyDict_GetItemRef().
  * PEP 703 proposes ``PyDict_FetchItem()`` name.
  * The majority seems to be fine with the proposed name: **PyDict_GetItemRef**().

* Type of the first parameter

  * My first version used ``PyObject*``
  * @markshannon wants ``PyDictObject*``, he created discussion: https://github.com/capi-workgroup/problems/issues/31
  * I changed the parameter type to ``PyDictObject*`` to see the consequences of the change: replace runtime check with assertion, remove tests on invalid dict type (since it would crash rather than raising a clean SystemError exception).
  * @serhiy-storchaka and @gpshead want ``PyObject*``: they are against the need for cast between ``PyObject*`` and ``PyDictObject*`` and do prefer type checking at runtime, since it's too rare that developers have access to a Python binary with assertions enabled
  * After @serhiy-storchaka and @gpshead pushback: I reverted this change, to come back to my initial plan: ``PyObject*``.

* @markshannon would prefer to define some guidelines first, before proposing adding a concrete function:

  *  "I'm opposed because making ad-hoc changes like this is going to make the C-API worse, not better."

* Since @markshannon has some disagreements, @gpshead opened a discussion at the Steering Council https://github.com/python/steering-council/issues/201 

---

PR changelog:

* First version
* Return 1 if the key is present
* Change parameter name from *pvalue* to *result* (and simplify dictitems_contains() implementation)
* Change first parameter type to ``PyDictObject**``
* Move the comments from C (dictobject.c) to header file (dictobject.h)
* Revert the parameter type change: come back to ``PyObject*``

**Review by serhiy-storchaka — APPROVED** [review]
I did not propose `PyDict_GetOptionalItem()` seriously, it would look confusing taking into account that other functions also return an "optional" item. I accept `PyDict_GetItemRef()` as token name and will accept any name whichever you choose. No need to bakeshed it to death.

I would be fine with the absent of runtime type check, not every C API function has a runtime type check, but the existing functions `PyDict_GetItem()` and `PyDict_GetItemWithError()`  have a runtime type check, so why this function should be special? And compile time type check would just not work.

I confirm my approve.

**Review by erlend-aasland — DISMISSED** [review]
Confirming my approval.

I believe introducing more type structs in the Limited API is a very bad decision, so I'm glad that particular change was reverted; I prefer to stick with the existing API convention of `PyObject *` only. For the "Brand New API", whatever that will be, a generic "PyObject-like struct" might not be the best option.

Regarding naming, I have no preference. Personally, I think I would have considered the PEP 703 naming to try and create slightly less merge conflict havoc for Sam ~~when~~ if nogil is going to be merged.

**Comment by encukou** [comment]
FWIW, here's a possible new variant: you could set `result` to `NULL` in which case the result isn't stored/incref'd. And that would start a convention of how to turn a get operation into a membership test. (And the `Lookup` name would fit that better.)

**Comment by markshannon** [comment]
If this function is to take `PyObject *`, as @erlend-aasland seems to insist, then it shouldn't raise a `SystemError` when passed something other than a dict. It should raise a `TypeError`.

**Comment by serhiy-storchaka** [comment]
`PyDict_GetItem()` and `PyDict_GetItemWithError()` raise a SystemError. All other functions in the Concrete Objects Layer either raise SystemError or have an undefined behavior. It is not specified what the behavior of the concrete function is, but passing an object of correct type is a requirement. https://docs.python.org/3/c-api/concrete.html

**Comment by vstinner** [comment]
I like SystemError. I used it in the past to detect bugs in C extensions, since TypeError is too common, so it was easy to distinguish "normal" TypeError from "invalid" SystemError. For me, SystemError means "bad API usage".

> If this function is to take PyObject *, as @erlend-aasland seems to insist, then it shouldn't raise a SystemError when passed something other than a dict. It should raise a TypeError.

IMO if TypeError is preferred, we should try to keep the whole C API consistent and raise TypeError in all similar situations, not only for newly added functions. Otherwise, it can be misleading.

**Comment by vstinner** [comment]
> FWIW, here's a possible new variant: you could set result to NULL in which case the result isn't stored/incref'd. And that would start a convention of how to turn a get operation into a membership test. (And the Lookup name would fit that better.)

For me, it's very surprising that the purpose of a function change "that much" depending on a parameter. For example, in Python, getattr() and hasattr() are two different functions: you cannot pass None to getattr() to check if an attribute exists.

I prefer to have separated functions to "get" an item/attribute and to check if an object "has" an item/attribute.

I dislike having to use locale.**set**locale(loc, None) to **get** a locale :-( I don't want to use locale.getlocale(loc) since it returns different different.

---

This issue goes in all directions, it's a little bit hard to follow :-(

**Comment by serhiy-storchaka** [comment]
> FWIW, here's a possible new variant: you could set result to NULL in which case the result isn't stored/incref'd.

How does this differ from `PyDict_Contains()`?

**Comment by vstinner** [comment]
Can we now please focus the discussion on this PR about PyDict_GetItemRef() and only PyDict_GetItemRef()?

For general C API discussions, I suggest using the https://github.com/capi-workgroup/problems/issues project.

So far, the PR got 2 approvals (@gpshead removed his approval).

Now I don't know if I'm supposed for wait for the SC: @gpshead created https://github.com/python/steering-council/issues/201 issue but in 2 weeks, the SC didn't have time to look into this issue.

**Comment by gpshead** [comment]
my 2 cents on exception types: If it is a bug in code at the C API level, `SystemError` makes sense.  If it is a bug in code at the Python level, `TypeError` makes sense.  Somewhat of a coin toss on some C APIs when the input in question can have come directly from the Python level (turning the question into one of who is responsible for pre-checking the type). This is also the kind of thing we can change in the future when doing a cleanup sweep of API behaviors to change some `SystemError`s coming from `PyErr_BadInternalCall` into TypeError as no code is _ever_ expected to catch and specifically handle SystemError.

For now, lets let this PR sit while we wait for steering council answers.

**Comment by vstinner** [comment]
> my 2 cents on exception types: If it is a bug in code at the C API level, SystemError makes sense. If it is a bug in code at the Python level, TypeError makes sense. Somewhat of a coin toss on some C APIs when the input in question can have come directly from the Python level (turning the question into one of who is responsible for pre-checking the type). This is also the kind of thing we can change in the future when doing a cleanup sweep of API behaviors to change some SystemErrors coming from PyErr_BadInternalCall into TypeError as no code is ever expected to catch and specifically handle SystemError.

PyObject_GetItem() is the safe **generic** API: it raises **TypeError** if the API is *misused*, if the first argument is not a mapping (or a sequence).

PyDict_GetItem() is the fast **specialized** API: it raises **SystemError** if it's *misused*, if the first argument is not a dict or a dict subtype.

**Comment by vstinner** [comment]
I rebased my PR on the main branch to fix conflicts.

**Comment by vstinner** [comment]
Thanks everyone for the great reviews. I think that the final API is better than my first version.

The road for that function was more bumpy than for other functions since PyDict_GetDict() is one of the commonly used function, and as I wrote before, we took this function as an opportunity to revisit some API design choices. There are some disagreements which have been discussed in length, especially in the https://github.com/capi-workgroup/problems/issues/ project. Overall, the majority seems to be in favor of this change (and I didn't see any concrete counter-proposal to solve the issue).

Sadly, this PR lost two approvals in the bumpy discussion. IMO it's now time to move on and see how this function can be used to avoid bugs and how to migrate users from the cursed PyDict_GetItem() to that better PyDict_GetItemRef() function.

Supporters of a new API instead of fixing the current API one function by one, I suggest continuing the discussion in https://github.com/capi-workgroup/problems/issues/ : there are a few issues related to that. So far, I didn't see any clear nor complete proposal, so for me, we are still at the same status quo: we do our best, fixing functions, one by one, when we agree that there is an issue. Same for the question of using ``PyDictObject*`` type rather than ``PyObject*`` for the first parameter: it's still being discussed and so far, no consensus was reached.

Completely getting rid of PyDict_GetItem() may take time. Maybe we need to consider https://github.com/capi-workgroup/problems/issues/54 approach for users who want to start a new C API with a clean C API without known issues like borrowed references. But well, that's out of the scope of this issue. This issue does **not** deprecate PyDict_GetItem() on purpose.

**Comment by vstinner** [comment]
I added PyDict_GetItemRef() to pythoncapi-compat: https://github.com/python/pythoncapi-compat/commit/eaff3c172f94ed32ac38860c38d7a8fa27483e57

**Pull request merged**
